src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    currentState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      for (const key in action.extraData) {
+        currentState[key] = action.extraData[key];
+      }
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    } else if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    history.push({ ...currentState });
```

Comment:
You are creating a new object copy of `currentState` in every iteration, which is correct to avoid mutation of the original state, but this action is repeated in every case. To adhere to the DRY principle, you should perform this action once after the `switch` statement.